homeassistant/components/switch/tellstick.py

```diff
@@ -4,98 +4,64 @@
-def setup_platform(hass, config, add_devices_callback, discovery_info=None):
+def setup_platform(hass, config, add_devices, discovery_info=None):
     """Setup Tellstick switches."""
-    import tellcore.telldus as telldus
-    import tellcore.constants as tellcore_constants
-    from tellcore.library import DirectCallbackDispatcher
-
-    core = telldus.TelldusCore(callback_dispatcher=DirectCallbackDispatcher())
-    signal_repetitions = config.get('signal_repetitions', SIGNAL_REPETITIONS)
-    switches_and_lights = core.devices()
-
-    switches = []
-    for switch in switches_and_lights:
-        if not switch.methods(tellcore_constants.TELLSTICK_DIM):
-            switches.append(
-                TellstickSwitchDevice(switch, signal_repetitions))
-
-    def _device_event_callback(id_, method, data, cid):
-        """Called from the TelldusCore library to update one device."""
-        for switch_device in switches:
-            if switch_device.tellstick_device.id == id_:
-                switch_device.update_ha_state()
-                break
-
-    callback_id = core.register_device_event(_device_event_callback)
+    if (discovery_info is None or
+            discovery_info[DISCOVER_DEVICES] is None or
+            TELLCORE_REGISTRY is None):
+        return
 
-    def unload_telldus_lib(event):
-        """Un-register the callback bindings."""
-        if callback_id is not None:
-            core.unregister_callback(callback_id)
+    # Allow platform level override, fallback to module config
+    signal_repetitions = discovery_info.get(DISCOVER_CONFIG,
+                                            SIGNAL_REPETITIONS)
 
-    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, unload_telldus_lib)
+    add_devices(TellstickSwitchDevice(
+        TELLCORE_REGISTRY.get_device(switch_id), signal_repetitions)
+                for switch_id in discovery_info[DISCOVER_DEVICES])
 
-    add_devices_callback(switches)
 
-
-class TellstickSwitchDevice(ToggleEntity):
+class TellstickSwitchDevice(TellstickDevice, ToggleEntity):
     """Representation of a Tellstick switch."""
 
     def __init__(self, tellstick_device, signal_repetitions):
         """Initialize the Tellstick switch."""
-        import tellcore.constants as tellcore_constants
-
-        self.tellstick_device = tellstick_device
-        self.signal_repetitions = signal_repetitions
-
-        self.last_sent_command_mask = (tellcore_constants.TELLSTICK_TURNON |
-                                       tellcore_constants.TELLSTICK_TURNOFF)
-
-    @property
-    def should_poll(self):
-        """No polling needed."""
-        return False
-
-    @property
-    def assumed_state(self):
-        """The Tellstick devices are always assumed state."""
-        return True
-
-    @property
-    def name(self):
-        """Return the name of the switch if any."""
-        return self.tellstick_device.name
+        TellstickDevice.__init__(self, tellstick_device, signal_repetitions)
```

Comment:
I am not sure if you actually need to overwrite the constructor if you're not changing anything.